inst/include/dust/dust.hpp

```diff
@@ -810,16 +858,44 @@ class Dust {
       }
       initialise_device_data();
     }
+    update_device_shared();
     _stale_host = false;
     _stale_device = true;
   }
 
   // This only gets called on construction; the size of these never
   // changes.
   void initialise_device_data() {
-    const size_t n_int = dust::device_work_size_int<T>(_shared[0]);
-    const size_t n_real = dust::device_work_size_real<T>(_shared[0]);
-    _device_data.initialise(_particles.size(), n_state_full(), n_int, n_real);
+    const auto s = _shared[0];
+    const size_t n_internal_int = dust::device_internal_size_int<T>(s);
+    const size_t n_internal_real = dust::device_internal_size_real<T>(s);
+    const size_t n_shared_int = dust::device_shared_size_int<T>(s);
+    const size_t n_shared_real = dust::device_shared_size_real<T>(s);
+    _device_data.initialise(_particles.size(), n_state_full(), _shared.size(),
+                            n_internal_int, n_internal_real,
+                            n_shared_int, n_shared_real);
+  }
+
+  template <typename U = T>
+  typename std::enable_if<!dust::has_gpu_support<U>::value, void>::type
+  update_device_shared() {
+  }
+
+  template <typename U = T>
+  typename std::enable_if<dust::has_gpu_support<U>::value, void>::type
+  update_device_shared() {
+    const size_t n_shared_int = _device_data.n_shared_int;
+    const size_t n_shared_real = _device_data.n_shared_real;
+    for (size_t i = 0; i < _shared.size(); ++i) {
+      // TODO: John, I've not used set_array here at least for now;
+      // this should work fine for the CPU but might need a bit of
+      // support on the GPU? If needed we can stage the copy on the
+      // cpu by putting it into one std::vector and then doing two big
+      // set_array calls. We never need to worry about the return leg.
```

Comment:
great minds (cf above comment)

Reply:
I think it probably depends on how often it gets called, we might get away with it if it's just done at the start. But thinking about the carehomes model (did you say about 1000 items in here?) maybe that will become a problem, especially if we start writing this more frequently, which seems possible with some inference methods

Reply:
An argument for doing this now is that we found profiling very difficult when we had loads of CUDA API calls

Reply:
OK, I've added a staged copy here, which should make the cuda work easier (1dc743e)